jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java

```diff
@@ -0,0 +1,284 @@
+        exitListener.countDown();
+        assertFalse(clientSocket.openLatch.await(2, TimeUnit.SECONDS));
+
+        // Strange that onError() is called but onOpen() is never called. (This would not be the case in Jetty-10).
+        assertThat(clientSocket.error.get(), instanceOf(CancellationException.class));
+    }
+
+    @Test
+    public void testAbortOnOpened() throws Exception
+    {
+        start(c -> c.addMapping("/", EventSocket.EchoSocket.class));
+
+        final CountDownLatch exitOnOpen = new CountDownLatch(1);
+        CloseTrackingEndpoint clientSocket = new CloseTrackingEndpoint()
+        {
+            @Override
+            public void onWebSocketConnect(Session session)
+            {
+                try
+                {
+                    super.onWebSocketConnect(session);
+                    exitOnOpen.await();
+                }
+                catch (InterruptedException e)
+                {
+                    throw new IllegalStateException(e);
+                }
+            }
+        };
+
+        // Abort during the call to onOpened. This is after future has been added to session,
+        // and after connection has been upgraded, but before future completion.
+        Future<Session> connect = client.connect(clientSocket, WSURI.toWebsocket(server.getURI()));
+        assertTrue(clientSocket.openLatch.await(5, TimeUnit.SECONDS));
+        assertTrue(connect.cancel(true));
+        exitOnOpen.countDown();
+
+        // We got an error on the WebSocket endpoint and an error from the future.
+        assertTrue(clientSocket.errorLatch.await(5, TimeUnit.SECONDS));
+        assertThrows(CancellationException.class, () -> connect.get(5, TimeUnit.SECONDS));
+    }
+
+    @Test
+    public void testAbortAfterCompletion() throws Exception
+    {
+        start(c -> c.addMapping("/", EventSocket.EchoSocket.class));
+
+        CloseTrackingEndpoint clientSocket = new CloseTrackingEndpoint();
+        Future<Session> connect = client.connect(clientSocket, WSURI.toWebsocket(server.getURI()));
+        Session session = connect.get(5, TimeUnit.SECONDS);
+
+        // If we can send and receive messages the future has been completed.
+        assertTrue(clientSocket.openLatch.await(5, TimeUnit.SECONDS));
+        clientSocket.getSession().getRemote().sendString("hello");
+        assertThat(clientSocket.messageQueue.poll(5, TimeUnit.SECONDS), Matchers.is("hello"));
+
+        // After it has been completed we should not get any errors from cancelling it.
+        assertFalse(connect.cancel(true));
+        assertThat(connect.get(5, TimeUnit.SECONDS), instanceOf(Session.class));
+        assertFalse(clientSocket.closeLatch.await(2, TimeUnit.SECONDS));
+        assertNull(clientSocket.error.get());
+
+        // Close the session properly.
+        session.close();
+        assertTrue(clientSocket.closeLatch.await(5, TimeUnit.SECONDS));
+        assertThat(clientSocket.closeCode, is(StatusCode.NORMAL));
+    }
+
```

Comment:
Please add a test case for the timeout of the future, i.e. calling `Future.get(1, SECONDS)`.
It should not fail the `connect()` since the semantic is just that of "1 second has passed" when it returns.
The test should then issue a second `Future.get()` that must succeed.

Also add a test that explicitly fails the future with a `CancellationException`, which should behave exactly like calling `Future.cancel()`.

Reply:
I wrote the first test case.

For the second one there is no way to fail the future with a specific `Exception`. Unless we downcast to `CompletableFuture<Session>`.

Reply:
Please downcast.